internal/runtime/kubernetes/kubeobserver/servicewaiter.go

```diff
@@ -59,13 +67,26 @@ func (w *serviceWaiter) Poll(ctx context.Context, c *k8s.Clientset) (bool, error
 		return false, err
 	}
 
+	id, ok := service.Labels[kubedef.K8sServerId]
+	if !ok {
+		return false, fnerrors.InternalError("service %q is missing server label", w.name)
+	}
+
+	pod, err := c.CoreV1().Pods(w.namespace).List(ctx, v1.ListOptions{
+		LabelSelector: kubedef.SerializeSelector(map[string]string{
+			kubedef.K8sServerId: id,
+		}),
+	})
+	if err != nil {
+		return false, err
+	}
+
 	var count int
 	for _, port := range service.Spec.Ports {
-		addr := fmt.Sprintf("%s.%s.svc.cluster.local:%d", service.Name, service.Namespace, port.Port)
+		addr := fmt.Sprintf("%s.%s.svc.cluster.local:%d", w.name, w.namespace, port.Port)
 
 		rawConn, err := net.DialTimeout("tcp", addr, dialTimeout)
-		if err != nil {
-			fmt.Fprintf(console.Debug(ctx), "failed to dial %s: %v\n", addr, err)
+		if !w.isReady(pod.Items, err) {
```

Comment:
I'm a bit confused about this. Ideally we wouldn't have to look for readiness at every step, but we would instead use the graph dependencies for that purpose. Why is checking for readiness required here?

Reply:
yeah, this is a bit messy. I mostly want to have a callback to emit events which can then trigger pod diagnostics if the service is not ready for a long time.
We could also just emit the dial error for now. It just felt like this will not be very helpful and us understanding which pods should serve this service, felt like a real benefit.